lonestar/CMakeLists.txt

```diff
@@ -23,6 +23,36 @@ function(add_test_scale type app)
   endforeach()
 endfunction(add_test_scale)
 
+function(app_cpu PREFIX name bin_name)
+  string(CONCAT target_name ${PREFIX} ${bin_name})
+  add_executable(${target_name} ${name}.cpp)
+  add_dependencies(apps ${target_name})
+  target_link_libraries(${target_name} PRIVATE Galois::shmem lonestar)
+  install(TARGETS ${target_name} DESTINATION "${CMAKE_INSTALL_BINDIR}" COMPONENT apps EXCLUDE_FROM_ALL)
+  set_target_properties(${target_name} PROPERTIES OUTPUT_NAME ${bin_name})
```

Comment:
I see. The first motivation of this work is that we want to create multiple binaries having the same name within the same CMake project. Now, it seems like we cannot avoid prefix, then.
For example, the analytic cpu tc, the analytic dist tc, and the mining cpu tc would be 'analy-cpu-triangle-counting', 'analy-dist-triangle-counting', 'mining-cpu-triangle-counting.'

What do you think about this? Do you have any advice on this? Thank you.

Reply:
I would a prefix a unique target/binary name in the entire project. It will keep things simple. 

Instead of a prefix, I would suggest a suffix. For example, triangle-counting-cpu, triangle-counting-gpu, triangle-counting-dist. I would also suggest to use these suffixes consistently for all apps. 
I would suggest to NOT use suffix for "type" of the app unless required. In the other words, do not use "analytics" and only use "mining" when you have to (that is, for triangle-counting-mining-cpu and triangle-counting-mining-gpu).

Reply:
I favor this too, though it contradicts what we discussed on Thursday. I was under the impression that we wanted to make the executable naming the same for the different variants to avoid inconveniencing the people developing the distributed/GPU versions. If this sounds better to everyone, let's do it this way. It will likely be easier to manage in the long run. Unfortunately it'll require quite more changes from @nicelhc13 here before we merge though.

One small nitpick with the naming. Is the cpu suffix really needed? Could we just do `triangle-counting` and `triangle-counting-dist`?

Reply:
There's triangle-counting-gpu as well. We can say the default (without suffix) is "cpu" but I think that add some overhead for the user. It's better to be explicit for the architecture in my opinion. Analytics/mining - there it's fine to omit where possible.

Reply:
Leaving the cpu suffix is also fine with me. With or without it, this is still by far the best naming scheme we've come up with thus far.

Reply:
I have one more question! To sum up, our rule is like below, to my best understanding.
1. Specify architecture type through an additional suffix (e.g. *-cpu, *-dist, *-gpu). In this case, dist attached apps could use either cpu or gpu, or both of them.

2. Omit the application types as many as possible. (It is a confusing part for me)
Based on the rule which shall omit 'analytic' suffix, do you also want to remove a suffix if the app is unique across the lonestar? For example, let's assume that lonestar/scientific/cpu only has delaunarytriangulation. In this case, that would be the 'delaunarytriangulation-cpu', not the 'delaunarytriangulation-scientific-cpu.'

Am I correct?

Reply:
It's easier if I list what I would prefer as suffix for each directory:

- lonestar/analytics/cpu : *-cpu
- lonestar/analytics/gpu : *-cpu
- lonestar/analytics/distributed : *-dist
- lonestar/eda/cpu : *-cpu 
- lonestar/mining/cpu : *-cpu (use *-mining-cpu only if there are conflicts)
- lonestar/mining/gpu : *-gpu (use *-mining-gpu only if there are conflicts)
- lonestar/scientific/cpu : *-cpu
- lonestar/scientifc/gpu : *-gpu

Reply:
> * lonestar/analytics/gpu : *-cpu
I think you meant `*-gpu` here. Other than that, yes, this is right.

And, just to clarify, the suffixes are not for the app directory names. That info is already present in the directory names and we don't need to add it twice.

Reply:
Yes! Apps in each directory, not the directory itself. Thanks.

Reply:
Cool. Thanks